notebooks/byod.py

```diff
@@ -98,40 +100,65 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        sample, ext = filename.rsplit(".", 1)
+        if sample.endswith(".cram"):
+            # double extension in the crai file, ie "foo.cram.crai"
+            sample, _ = sample.rsplit(".", 1)
+        return sample, ext
+
     def create_cram_crai_table(table: str, listing: Iterable[str]):
-        crams = dict()
-        crais = dict()
+        matched_files = defaultdict(dict)  # type: dict
+        allCrams = []
+        allCrais = []
         for key in listing:
             _, filename = key.rsplit("/", 1)
-
-            parts = filename.split(".")
-            if 3 == len(parts):  # foo.cram.crai branch
-                sample, _, ext = parts
-            elif 2 == len(parts):  # "foo.cram" or "foo.crai" branch
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
-            if "cram" == ext:
-                crams[sample] = key
-            elif "crai" == ext:
-                crais[sample] = key
-            else:
-                continue
-        samples = sorted(crams.keys())
-        upload_columns(table, dict(sample=samples,
-                                   cram=[crams[s] for s in samples],
-                                   crai=[crais[s] for s in samples]))
+            sample, ext = parse_cram_crai(filename)
+            if ext in ['cram', 'crai']:
+                matched_files[sample][ext] = filename
+                if ext == 'cram':
+                    allCrams.append(sample)
+                else:
+                    allCrais.append(sample)
+
+        if allCrams != allCrais:
+            # We don't want to iterate an additional time if we don't need to,
+            # so this only runs if there's more crams than crais or vice versa
+            for sample in list(matched_files):
+                try:
+                    if matched_files[sample]['cram'] and not matched_files[sample]['crai']:
```

Comment:
Will:

```
                    if matched_files[sample]['cram'] and not matched_files[sample]['crai']:
                        if not matched_files[sample]['cram'] and matched_files[sample]['crai']:
```

ever occur?

Reply:
Either one of those lines can throw a KeyError, yes. It'll be the first line if there is no crai and the second line if there is no cram. They're nested because if they were not, this happened:
fizz is missing a crai --> keyerror, deletes fizz --> fizz is missing a cram --> keyerror again
Instead of...
fizz is missing a crai --> keyerror, deletes fizz --> moves on

Oddly enough I didn't get KeyErrors earlier on; you'll see older commits handled this better because without KeyErrors I can account for either case. I think something is wrong with my defaultdictionary setup; I had to go from defaultdict(dict()) to defaultdict() at some point.

Reply:
Have you heard of `dict().get(item)`?  As it is, this code is an impossible conditional.

Reply:
Found a solution. It's not exactly elegant, though. --> https://github.com/DataBiosphere/bdcat_notebooks/pull/27/commits/93f0b4a780db92fcda3e982bd4943c78f2e5a0da